galadriel/proof/prover.py

```
@@ -0,0 +1,89 @@
+    def __init__(self):
+        with open(PRIVATE_KEY_PATH, "rb") as priv_file:
+            self.private_key = serialization.load_pem_private_key(data=priv_file.read(), password=None)
+
+        with open(PUBLIC_KEY_PATH, "rb") as pub_file:
+            self.public_key = serialization.load_pem_public_key(data=pub_file.read())
+            self.public_key_bytes = self.public_key.public_bytes(
+                encoding=serialization.Encoding.Raw, format=serialization.PublicFormat.Raw
+            )
+        self.nsm_util = NSMUtil()
+
+    async def generate_proof(self, request: Message, response: Message) -> Proof:
+        try:
+            # Hash data
+            hashed_data = self._hash_data(request, response)
+
+            # Sign data directly using the private key
+            signature = self.private_key.sign(hashed_data)
+
+            # Get attestation doc
+            attestation_doc = self.nsm_util.get_attestation_doc(self.public_key_bytes)
+
+            proof = Proof(
+                hash=hashed_data.hex(),
+                signature=signature.hex(),
+                public_key=self.public_key_bytes.hex(),
+                attestation=base64.b64encode(attestation_doc).decode(),
+            )
+        except Exception as e:
+            logger.error(f"Error generating proof: {e}")
+            raise e
+
+        return proof
+
+    def _hash_data(self, request: Message, response: Message) -> bytes:
+        combined_str = (
+            f"{json.dumps(request.model_dump(), sort_keys=True)}{json.dumps(response.model_dump(), sort_keys=True)}"
+        )
+        return hashlib.sha256(combined_str.encode("utf-8")).digest()
+
+    async def publish_proof(self, request: Message, response: Message, proof: Proof) -> bool:
+        url = "https://api.galadriel.com/v1/verified/chat/log"
+        headers = {
+            "accept": "application/json",
+            "Content-Type": "application/json",
+            "Authorization": self._get_authorization(),
```

imo would make more sense to check if API key present, if not fail early?